File: LinkedList/CircularyLinkedList/Java/CLL.java
class CLL<T>{

    private class Node{

        T value;
        Node next;

        public Node(T value){
            this.value = value;
            this.next = null;
        }
    }

    private Node head;
    private Node tail;
    private int size;

    public CLL(){
        head = tail = null;
        size = 0;
    }

    public void insertFirst(T value){        
        Node newNode = new Node(value);
        if(head == null){
            head = newNode;
            tail = newNode;
            newNode.next = head;
        }else{
            newNode.next = head;
            head = newNode;
            tail.next = head;
        }
        size++;      
    }

    public void insertLast(T value){
        if(head == null){
            insertFirst(value);
        }else{
            Node newNode = new Node(value);
            tail.next = newNode;
            tail = newNode;
            tail.next = head;
            size++;
        }
    }

    public void removeFirst(){
        if(head == null){
            return;
        }
        head = head.next;
        tail.next = head;
        size--; 
    }

    public void printList(){
        Node current = head;
        while(true){
            System.out.print(current.value + " ");
            current = current.next;
            if(current == head){
                break;
            }
        }
        System.out.println();
    }

    public int getSize(){
        return size;
    }   
}

File: LinkedList/CircularyLinkedList/C++/CLL.cpp
#include <iostream>

class Node{
    
    public:
        int value;
        Node *next;

        Node() = default;
};

class CLL{

    private:
        Node *head;
        Node *tail;
        int size;

    public:
        CLL(){
            head = tail = nullptr;
            size = 0;
        }

        void insertFirst(int value){
            Node *newNode = new Node;
            newNode->value = value;
            if(head == nullptr){
                head = tail = newNode;
                head->next = head;
            }else{
                newNode->next = head;
                head = newNode;
                tail->next = head;
            }
            size++;
        }

        void insertLast(int value){
            if(head == nullptr){
                insertFirst(value);
            }else{
                Node *newNode = new Node;
                newNode->value = value;
                tail->next = newNode;
                tail = newNode;
                tail->next = head;
                size++;
            }
        }

        void removeFirst(){
            if(head == nullptr){
                return;
            }
            if(head->next == head){
                head = nullptr;
                return;
            }
            head = head->next;
            tail->next = head;
            size--;
        }

        void printList(){
            if(head != nullptr){
                Node *current = head;
                while(true){
                    std::cout << current->value << " ";
                    current = current->next;
                    if(current == head){
                        break;
                    }
                }
            }            
            std::cout << std::endl;
        }

        int getSize(){
            return size;
        }     

};

int main(){
    CLL l;
    l.insertFirst(2);
    l.insertFirst(1);
    l.insertLast(3);
    l.removeFirst();
    l.removeFirst();
    l.removeFirst();
    l.removeFirst();
    l.printList();

    return 0;
}
